cpp/kernels_non_const_coefficient.hpp

```diff
@@ -0,0 +1,422 @@
+  //           const int* entity_local_index, const std::uint8_t* quadrature_permutation)
+  // {
+  //   assert(bs == 3);
+  //   // Get geometrical data
+  //   xt::xtensor<double, 2> J = xt::zeros<double>({gdim, tdim});
+  //   xt::xtensor<double, 2> K = xt::zeros<double>({tdim, gdim});
+  //   std::array<std::size_t, 2> shape = {d, gdim};
+  //   xt::xtensor<double, 2> coord = xt::adapt(coordinate_dofs, gdim * d, xt::no_ownership(),
+  //   shape);
+
+  //   // Compute Jacobian, its inverse and the determinant
+  //   dolfinx_cuas::math::compute_jacobian(dphi0_c, coord, J);
+  //   dolfinx_cuas::math::compute_inv(J, K);
+  //   double detJ = std::fabs(dolfinx_cuas::math::compute_determinant(J));
+
+  //   // Temporary variable for grad(phi) on physical cell
+  //   xt::xtensor<double, 2> dphi_phys({ndofs_cell, bs});
+
+  //   // Main loop
+  //   for (std::size_t q = 0; q < weights.size(); q++)
+  //   {
+  //     const double w0 = weights[q] * detJ;
+  //     // Precompute J^-T * dphi
+  //     std::fill(dphi_phys.begin(), dphi_phys.end(), 0);
+  //     for (int i = 0; i < ndofs_cell; i++)
+  //     {
+  //       for (int j = 0; j < bs; j++)
+  //       {
+  //         double acc = 0;
+  //         for (int k = 0; k < tdim; k++)
+  //           acc += K(k, j) * _dphi(q, i, k);
+  //         dphi_phys(i, j) += acc;
+  //       }
+  //     }
+
+  //     // Add contributions to local matrix
+  //     for (int i = 0; i < ndofs_cell; i++)
+  //     {
+  //       for (int j = 0; j < ndofs_cell; j++)
+  //       {
+  //         // Compute block invariant term from sigma(u):eps(v)
+  //         double block_invariant = 0;
+  //         for (int s = 0; s < bs; s++)
+  //           block_invariant += dphi_phys(i, s) * dphi_phys(j, s);
+  //         block_invariant *= w0;
+
+  //         for (int k = 0; k < bs; ++k)
+  //         {
+  //           const size_t row = (k + i * bs) * (ndofs_cell * bs);
+
+  //           // Add block invariant term from sigma(u):eps(v)
+  //           A[row + j * bs + k] += block_invariant;
+
+  //           for (int l = 0; l < bs; ++l)
+  //             A[row + j * bs + l] += dphi_phys(i, l) * dphi_phys(j, k) * w0;
+  //         }
+  //       }
+  //     }
+  //   }
+  // };
+
+  switch (type)
+  {
+  case dolfinx_cuas::Kernel::Mass:
+    return mass_coeff;
+  // case dolfinx_cuas::Kernel::MassTensor:
+  //   return masstensor;
+  // case dolfinx_cuas::Kernel::Stiffness:
+  //   return stiffness;
+  // case dolfinx_cuas::Kernel::TrEps:
+  //   return tr_eps;
+  // case dolfinx_cuas::Kernel::SymGrad:
+  //   return sym_grad_eps;
+  default:
+    throw std::runtime_error("unrecognized kernel");
+  }
+}
+} // namespace
+namespace dolfinx_cuas
+{
+
+/// Create integration kernel for Pth order Lagrange elements
+/// @param[in] type The kernel type (Mass or Stiffness)
+/// @param[in] P Degree of the element
+/// @param[in] bs The block size
+/// @return The integration kernel
+kernel_fn generate_coeff_kernel(
```

Comment:
Have you tested this with block size?
The kernels Igor implemented does not support it. I have a branch where this is about to change

Reply:
No. This is based on Igor's kernel. I can wait with merging and add your changes.